### PR TITLE
Steps to Care - Security Permissions and Notification Preferences

### LINKED ISSUE #83: B: Steps to Care Prayer, Benevolence, Connection Actions wording confusing
# Expected Behavior

On Care Dashboard, when selecting the options, "Convert to Prayer Request, Benevolence Request, or Connections Request" seems to suggest that the care need will be changed to one of the former.

# Current Behavior
  When selecting "Convert to Prayer Request, Benevolence Request, or Connections Request" it adds a prayer, benevolence, or connections request.  The verbiage might be clearer if we use the wording "Add prayer, benevolence or connections request."

Screen Shot 2021-11-02 at 11 48 39 AM

### PR DESCRIPTION
**Pull request by nateh777**
##### What does the change add or fix?

- Added ability to flag Need's as "Workers Only".
- Added additional security permissions for "View All" and "Care Workers".
- Added Notification preferences to the Care Dashboard.
- Added logic to Care Entry to be able to handle new notification preferences.
- Fixes #83 

**New Settings:**

Security Permissions

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

- Added ability to flag Need's as "Workers Only".
- Added additional security permissions for "View All" and "Care Workers".
- Added Notification preferences to the Care Dashboard.

---------

##### Who reported, requested, or paid for the change?

Rock Springs

---------

##### Does this update or add options to the block UI?

image

image

image

image

image

---------

##### What files does it affect?

- StepsToCare/CareDashboard.ascx
- StepsToCare/CareDashboard.ascx.cs
- StepsToCare/CareEntry.ascx
- StepsToCare/CareEntry.ascx.cs

---------

##### Is it a breaking change for other versions/clients?

Steps to Care now requires 12.4 due to an added GetAttributeValue method I am using to bypass looking up an attribute key by guid. Also requires new Assembly update in https://github.com/KingdomFirst/RockAssemblies/pull/51

### CODE REVIEW

File: StepsToCare/CareDashboard.ascx.cs [hunk 1]
```diff
@@ -368,23 +405,40 @@ protected override void OnInit( EventArgs e )
             rFilter.ApplyFilterClick += rFilter_ApplyFilterClick;
             rFollowUpFilter.ApplyFilterClick += rFollowUpFilter_ApplyFilterClick;
 
-            _canAddEditDelete = IsUserAuthorized( Authorization.EDIT );
+            _canEdit = IsUserAuthorized( Authorization.EDIT );
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:408]:
```suggestion
            _canEdit = IsUserAuthorized( Authorization.EDIT ) || IsUserAuthorized( Authorization.ADMINISTRATE );
```

File: StepsToCare/CareDashboard.ascx.cs [hunk 2]
```diff
@@ -1407,6 +1483,25 @@ protected void mdMakeNote_SaveClick( object sender, EventArgs e )
...
+        {
+            var rockContext = new RockContext();
+            var attribute = AttributeCache.Get( rocks.kfs.StepsToCare.SystemGuid.PersonAttribute.NOTIFICATION );
+
+            if ( CurrentPerson != null )
+            {
+                Control attributeControl = phNotificationAttribute.FindControl( string.Format( "attribute_field_{0}", attribute.Id ) );
+                if ( attributeControl != null )
+                {
+                    string originalValue = CurrentPerson.GetAttributeValue( attribute.Key );
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:1496]: Is the original value necessary?

File: StepsToCare/CareDashboard.ascx.cs [hunk 3]
```diff
@@ -2103,6 +2224,26 @@ private string ResolveValues( string values, CheckBoxList checkBoxList )
...
         }
 
+        private void AddNotificationAttributeControl( bool setValues )
+        {
+            phNotificationAttribute.Controls.Clear();
+
+            var attribute = AttributeCache.Get( rocks.kfs.StepsToCare.SystemGuid.PersonAttribute.NOTIFICATION );
+            CurrentPerson.LoadAttributes();
+            string attributeValue = CurrentPerson.GetAttributeValue( attribute.Key );
+            string attributeLabel = attribute.Name;
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:2234]:
```suggestion
```

File: StepsToCare/CareDashboard.ascx.cs [hunk 4]
```diff
@@ -2103,6 +2224,26 @@ private string ResolveValues( string values, CheckBoxList checkBoxList )
...
+            attribute.AddControl( phNotificationAttribute.Controls, attributeValue, "Notification", setValues, true, true, "Choose how you would like to receive Steps to Care notifications", "" );
+
+            if ( CurrentPerson.CanReceiveEmail( false ) )
+            {
+                nbNotificationWarning.Text = "Please ensure your email address is setup properly on your profile. ";
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:2239]:
```suggestion
                nbNotificationWarning.Text = "Please ensure your email address is setup properly on your profile. ";
                nbNotificationWarning.Visible = true;
```

File: StepsToCare/CareDashboard.ascx.cs [hunk 5]
```diff
@@ -2103,6 +2224,26 @@ private string ResolveValues( string values, CheckBoxList checkBoxList )
...
+            }
+            if ( CurrentPerson.PhoneNumbers.GetFirstSmsNumber() == "" )
+            {
+                nbNotificationWarning.Text += "Please ensure you have a valid phone number attached to your profile and SMS is allowed. ";
```
Comment (nateh777) [inline StepsToCare/CareDashboard.ascx.cs:2243]:
```suggestion
                nbNotificationWarning.Text += "Please ensure you have a valid phone number attached to your profile and SMS is allowed. ";
                nbNotificationWarning.Visible = true;
```

File: StepsToCare/CareEntry.ascx.cs [hunk 6]
```diff
@@ -402,26 +406,69 @@ protected void lbSave_Click( object sender, EventArgs e )
...
+                        var systemCommunication = new SystemCommunicationService( rockContext ).Get( assignmentEmailTemplateGuid.Value );
+                        var emailMessage = new RockEmailMessage( systemCommunication );
+                        var smsMessage = new RockSMSMessage( systemCommunication );
+                        emailMessage.AppRoot = smsMessage.AppRoot = ResolveRockUrl( "~/" );
+                        emailMessage.ThemeRoot = smsMessage.ThemeRoot = ResolveRockUrl( "~~/" );
                         foreach ( var assignee in assignedPersons )
                         {
+                            assignee.PersonAlias.Person.LoadAttributes();
+                            var smsNumber = assignee.PersonAlias.Person.PhoneNumbers.GetFirstSmsNumber();
+
```
Comment (nateh777) [inline StepsToCare/CareEntry.ascx.cs:426]: Add can receive email or empty sms number check here

File: StepsToCare/CareDashboard.ascx.cs (hunk 5 above)
Comment (nateh777) [reply StepsToCare/CareDashboard.ascx.cs]: Resolved it slightly differently when I found there were a few bugs

File: StepsToCare/CareDashboard.ascx.cs (hunk 4 above)
Comment (nateh777) [reply StepsToCare/CareDashboard.ascx.cs:2237]: Resolved it slightly differently when I found there were a few bugs

File: StepsToCare/CareDashboard.ascx.cs [hunk 7]
```diff
@@ -2231,16 +2230,21 @@ private void AddNotificationAttributeControl( bool setValues )
             var attribute = AttributeCache.Get( rocks.kfs.StepsToCare.SystemGuid.PersonAttribute.NOTIFICATION );
             CurrentPerson.LoadAttributes();
             string attributeValue = CurrentPerson.GetAttributeValue( attribute.Key );
-            string attributeLabel = attribute.Name;
             attribute.AddControl( phNotificationAttribute.Controls, attributeValue, "Notification", setValues, true, true, "Choose how you would like to receive Steps to Care notifications", "" );
 
-            if ( CurrentPerson.CanReceiveEmail( false ) )
+            if ( !CurrentPerson.CanReceiveEmail( false ) )
```
Comment (guymaness2) [inline StepsToCare/CareDashboard.ascx.cs:2235]: Can you help me understand this logic changing to a "not"?

**Review by guymaness2 — APPROVED** [review]

**Pull request merged**